File: code/redis/examples/sentinel_manager.go
package main

/*
Redis Sentinel 管理システム

使用方法:
    sentinelHosts := []string{
        "sentinel1.example.com:26379",
        "sentinel2.example.com:26379",
        "sentinel3.example.com:26379",
    }
    manager := NewRedisSentinelManager(sentinelHosts, "mymaster")
    master := manager.GetMaster()
*/

import (
	"context"
	"fmt"
	"log"
	"time"

	"github.com/go-redis/redis/v8"
)

var ctx = context.Background()

type RedisSentinelManager struct {
	sentinelClient *redis.Client
	serviceName    string
	masterClient   *redis.Client
	slaveClient    *redis.Client
}

func NewRedisSentinelManager(sentinelAddrs []string, serviceName string) *RedisSentinelManager {
	if serviceName == "" {
		serviceName = "mymaster"
	}

	sentinelClient := redis.NewFailoverClient(&redis.FailoverOptions{
		MasterName:    serviceName,
		SentinelAddrs: sentinelAddrs,
	})

	return &RedisSentinelManager{
		sentinelClient: sentinelClient,
		serviceName:    serviceName,
	}
}

func (m *RedisSentinelManager) GetMaster() *redis.Client {
	if m.masterClient == nil {
		m.masterClient = m.sentinelClient
	}
	return m.masterClient
}

func (m *RedisSentinelManager) GetSlave() *redis.Client {
	// 読み取り専用のスレーブ接続を取得
	// Note: go-redis/v8では、FailoverClientが自動的にマスター/スレーブを管理
	if m.slaveClient == nil {
		m.slaveClient = m.sentinelClient
	}
	return m.slaveClient
}

func (m *RedisSentinelManager) GetMasterInfo() (map[string]string, error) {
	result, err := m.sentinelClient.Info(ctx, "replication").Result()
	if err != nil {
		log.Printf("Master discovery failed: %v", err)
		return nil, err
	}

	info := make(map[string]string)
	info["replication_info"] = result
	return info, nil
}

func (m *RedisSentinelManager) WaitForMaster(timeout time.Duration) bool {
	deadline := time.Now().Add(timeout)

	for time.Now().Before(deadline) {
		if _, err := m.GetMasterInfo(); err == nil {
			log.Println("Master available")
			return true
		}
		time.Sleep(1 * time.Second)
	}

	log.Println("Master not available within timeout")
	return false
}

func WriteData(manager *RedisSentinelManager, key, value string) error {
	master := manager.GetMaster()
	return master.Set(ctx, key, value, 0).Err()
}

func ReadData(manager *RedisSentinelManager, key string) (string, error) {
	// 読み取りはスレーブから
	slave := manager.GetSlave()
	val, err := slave.Get(ctx, key).Result()
	if err == nil {
		return val, nil
	}

	// スレーブが利用できない場合はマスターから
	master := manager.GetMaster()
	return master.Get(ctx, key).Result()
}

func main() {
	sentinelHosts := []string{
		"sentinel1.example.com:26379",
		"sentinel2.example.com:26379",
		"sentinel3.example.com:26379",
	}

	manager := NewRedisSentinelManager(sentinelHosts, "mymaster")

	// 書き込み
	if err := WriteData(manager, "test_key", "test_value"); err != nil {
		log.Printf("書き込みエラー: %v", err)
	} else {
		fmt.Println("データ書き込み成功")
	}

	// 読み取り
	value, err := ReadData(manager, "test_key")
	if err != nil {
		log.Printf("読み取りエラー: %v", err)
	} else {
		fmt.Printf("読み取り値: %s\n", value)
	}

	// マスター情報取得
	info, err := manager.GetMasterInfo()
	if err != nil {
		log.Printf("マスター情報取得エラー: %v", err)
	} else {
		fmt.Printf("マスター情報: %v\n", info)
	}
}

File: code/redis/examples/sentinel_manager.py
#!/usr/bin/env python3
"""
Redis Sentinel 管理システム

使用方法:
    from sentinel_manager import RedisSentinelManager
    
    sentinel_hosts = [
        ('sentinel1.example.com', 26379),
        ('sentinel2.example.com', 26379),
        ('sentinel3.example.com', 26379)
    ]
    
    sentinel_manager = RedisSentinelManager(sentinel_hosts)
    master = sentinel_manager.get_master()
"""

import redis.sentinel
import logging
import time
from typing import List, Optional

class RedisSentinelManager:
    def __init__(self, sentinel_hosts: List[tuple], service_name: str = 'mymaster'):
        """
        Redis Sentinel接続管理
        
        Args:
            sentinel_hosts: [(host, port), ...] のリスト
            service_name: Sentinelで管理されるサービス名
        """
        self.sentinel_hosts = sentinel_hosts
        self.service_name = service_name
        self.sentinel = redis.sentinel.Sentinel(sentinel_hosts)
        self.logger = logging.getLogger(__name__)
    
    def get_master(self):
        """マスターRedis接続取得"""
        try:
            master = self.sentinel.master_for(
                self.service_name,
                socket_timeout=0.1,
                password=None,
                db=0
            )
            return master
        except Exception as e:
            self.logger.error(f"Master connection failed: {e}")
            return None
    
    def get_slave(self):
        """スレーブRedis接続取得（読み取り専用）"""
        try:
            slave = self.sentinel.slave_for(
                self.service_name,
                socket_timeout=0.1,
                password=None,
                db=0
            )
            return slave
        except Exception as e:
            self.logger.error(f"Slave connection failed: {e}")
            return None
    
    def get_master_info(self):
        """マスター情報取得"""
        try:
            return self.sentinel.discover_master(self.service_name)
        except Exception as e:
            self.logger.error(f"Master discovery failed: {e}")
            return None
    
    def get_slaves_info(self):
        """スレーブ情報取得"""
        try:
            return self.sentinel.discover_slaves(self.service_name)
        except Exception as e:
            self.logger.error(f"Slaves discovery failed: {e}")
            return []
    
    def wait_for_master(self, timeout: int = 30):
        """マスター復旧待機"""
        start_time = time.time()
        while time.time() - start_time < timeout:
            master_info = self.get_master_info()
            if master_info:
                self.logger.info(f"Master available: {master_info}")
                return True
            time.sleep(1)
        
        self.logger.error("Master not available within timeout")
        return False

# 使用例
if __name__ == "__main__":
    sentinel_hosts = [
        ('sentinel1.example.com', 26379),
        ('sentinel2.example.com', 26379),
        ('sentinel3.example.com', 26379)
    ]
    
    sentinel_manager = RedisSentinelManager(sentinel_hosts)
    
    # 読み書き分離
    def write_data(key, value):
        master = sentinel_manager.get_master()
        if master:
            return master.set(key, value)
        return False
    
    def read_data(key):
        # 読み取りはスレーブから
        slave = sentinel_manager.get_slave()
        if slave:
            return slave.get(key)
        
        # スレーブが利用できない場合はマスターから
        master = sentinel_manager.get_master()
        if master:
            return master.get(key)
        
        return None
